ethereum-rpc-co/src/main/java/pm/gnosis/ethereum/rpc/co/RpcEthereumRepository.kt

```diff
@@ -0,0 +1,135 @@
+package pm.gnosis.ethereum.rpc.co
+
+import pm.gnosis.ethereum.*
+import pm.gnosis.ethereum.models.EthereumBlock
+import pm.gnosis.ethereum.models.TransactionData
+import pm.gnosis.ethereum.models.TransactionParameters
+import pm.gnosis.ethereum.models.TransactionReceipt
+import pm.gnosis.ethereum.rpc.EthereumRpcConnector
+import pm.gnosis.ethereum.rpc.models.*
+import pm.gnosis.ethereum.rpc.toRpcRequest
+import pm.gnosis.model.Solidity
+import pm.gnosis.models.Transaction
+import pm.gnosis.models.Wei
+import java.math.BigDecimal
+
+class RpcEthereumRepository(
+    private val ethereumRpcApi: EthereumRpcConnector
+) : EthereumRepository {
+
+    override suspend fun <R : BulkRequest> request(bulk: R): R =
+        bulk.requests.associate { it.id to it.toRpcRequest() }
+            .let { rpcRequests ->
+                ethereumRpcApi.post(rpcRequests.values.map { it.request() })
+                    .map { rpcRequests[it.id]?.parse(it) }
+                    .let { bulk }
+            }
+
+    override suspend fun <R : EthRequest<*>> request(request: R): R =
+        request.toRpcRequest().let { rpcRequest ->
+            ethereumRpcApi.post(rpcRequest.request())
+                .let {
+                    rpcRequest.parse(it)
+                    request
+                }
+        }
+
+    override suspend fun getBalance(address: Solidity.Address): Wei =
+        request(EthBalance(address)).checkedResult()
+
+    override suspend fun sendRawTransaction(signedTransactionData: String): String =
+        request(EthSendRawTransaction(signedTransactionData))
+            .checkedResult("Could not send raw transaction")
+
+    override suspend fun getTransactionReceipt(transactionHash: String): TransactionReceipt =
+        ethereumRpcApi.receipt(
+            JsonRpcRequest(
+                method = "eth_getTransactionReceipt",
+                params = listOf(transactionHash)
+            )
+        ).result?.let {
+            TransactionReceipt(
+                it.status,
+                it.transactionHash,
+                it.transactionIndex,
+                it.blockHash,
+                it.blockNumber,
+                it.from,
+                it.to,
+                it.cumulativeGasUsed,
+                it.gasUsed,
+                it.contractAddress,
+                it.logs.map { TransactionReceipt.Event(it.logIndex, it.data, it.topics) }
+            )
+        } ?: throw TransactionReceiptNotFound()
+
+
+    override suspend fun getTransactionByHash(transactionHash: String): TransactionData =
```

Comment:
This feels weird, as we have the whole logic twice (the part that builds the JsonRpcRequest). That should only exist once, right?

so what I would do here is create a scope (and maybe allow configuring that), use the sync interface and wrap the call in the scope.

Or in the first run we say this is not something that needs be solved here and just provide the sync interface. 

OR we say ... coroutines is the way to go (and we want to use them properly with retrofit) so for now we do not provide anything else.

Reply:
I think it's worth considering supporting just coroutines. This is also the approach we use in the app. Having rx will create additional maintenance and tech debt and we don't know for sure if it is ever gonna be needed. Supporting both at the beginning will slow us down. We still can add support for rx at some later point.

Reply:
Actually, another argument in favour of having only coroutines is that there are already adapters implemented that will wrap everything in some form of rx stream: https://github.com/Kotlin/kotlinx.coroutines/tree/master/reactive/kotlinx-coroutines-rx2

Reply:
If we all agree then lets do that :D